### PR TITLE
PIM-10149: implementation with a GroupProductsCommand

### PR DESCRIPTION
**Pull request by rfauglas**
Risk analysis:
https://www.notion.so/akeneo/Analysis-of-OOM-problem-during-Product-Group-update-e0b79151a9df4575838aaa4c26259286

**Description (for Contributor and Core Developer)**

Bug is out of memory for product group edit page when many products are attached...

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [x] Migration & Installer
- [ ] PM Validation (Story)
- [x] Changelog (maintenance bug fixes)
- [x] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/GroupController.php [hunk 1]
```diff
@@ -195,6 +194,10 @@ public function postAction(Request $request, $code)
 
         $this->saver->save($group);
 
+        if (array_key_exists('products', $data)) {
+            $this->updateProductsToGroupHandler->handle(new UpdateProductsToGroupCommand($group->getId(), $data['products']));
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/GroupController.php:198]:
```suggestion
            $this->updateProductsToGroupHandler->handle(new UpdateProductsToGroupCommand($group->getCode(), $data['products']));
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    public function getGroupId(): string
+    {
+        return $this->groupId;
+    }
+
+    /**
+     * @return array
+     */
+    public function getUptodateProductIds(): array
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php:26]:
```suggestion
    public function productIds(): array
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ */
+class UpdateProductsToGroupCommand
+{
+    private string $groupId;
+    private array $uptodateProductIds;
+
+    /**
+     * @return string
+     */
+    public function getGroupId(): string
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php:18]:
```suggestion
    public function groupCode(): string
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+
+namespace Akeneo\Pim\Enrichment\Component\Product\Commands;
+
+/**
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class UpdateProductsToGroupCommand
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php:10]: `GroupProductCommand`? "update" is not appropriated as it's a user intention and not CRUD here.

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $this->getGroupProductIdentifiers = $getGroupProductIdentifiers;
+        $this->entityManager = $entityManager;
+        $this->productSaver = $productSaver;
+    }
+
+    public function handle(UpdateProductsToGroupCommand $updateProductsToGroupCommand)
+    {
+        $uptodateProductIds = $updateProductsToGroupCommand->getUptodateProductIds();
+        $oldProductIds = $this->getGroupProductIdentifiers->byGroupId($updateProductsToGroupCommand->getGroupId());
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php:32]: FindProductIdsInGroup->fromGroupCode

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    public function __construct(GetGroupProductIdentifiers $getGroupProductIdentifiers, EntityManager $entityManager, BulkSaverInterface $productSaver)
+    {
+        $this->getGroupProductIdentifiers = $getGroupProductIdentifiers;
+        $this->entityManager = $entityManager;
+        $this->productSaver = $productSaver;
+    }
+
+    public function handle(UpdateProductsToGroupCommand $updateProductsToGroupCommand)
+    {
+        $uptodateProductIds = $updateProductsToGroupCommand->getUptodateProductIds();
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php:31]:
```suggestion
        $currentProductIds = $updateProductsToGroupCommand->getUptodateProductIds();
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        $oldProductIds = $this->getGroupProductIdentifiers->byGroupId($updateProductsToGroupCommand->getGroupId());
+
+        $newProductIds = array_diff($uptodateProductIds, $oldProductIds);
+        $removedProductIds = array_diff($oldProductIds, $uptodateProductIds);
+
+        $productsToUpdate = [];
+        $group = $this->entityManager->find(Group::class, $updateProductsToGroupCommand->getGroupId());
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php:38]: inject directly the group repository

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+        foreach ($newProductIds as $newProductId) {
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php:40]: it will fail anyway in an HTTP request, as the issue is with 16k products...

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php (hunk 4 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php:10]: Yes you're fully right group as verb not an entity...

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php (hunk 5 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php:32]: I reused existing code...

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php (hunk 7 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php:38]: I understand your intention, but I thought it would make the comment less entity dependant...If we say that Command is a traduction from user intention... it also detached from any doctrine context...

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php (hunk 8 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php:40]: yes indeed but as you said, little chance to happen... that a user updates 16k products in a request...

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php (hunk 2 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php]: done!

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/GroupController.php (hunk 1 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/GroupController.php]: keep ID after discussion.

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php (hunk 3 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupCommand.php]: changed to groupId()

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php (hunk 5 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/UpdateProductsToGroupHandler.php]: changed tFindProductIdentifiersInterface::fromGroupId

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        $this->findGroupProductIdentifiers = $getGroupProductIdentifiers;
+        $this->groupRepository = $groupRepository;
+        $this->productSaver = $productSaver;
+        $this->productRepository = $productRepository;
+    }
+
+    public function handle(GroupProductsCommand $updateProductsToGroupCommand)
+    {
+        $currentProductIds = $updateProductsToGroupCommand->productIds();
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php:33]:
```suggestion
        $newProductIds = $updateProductsToGroupCommand->productIds();
```

current is confusing. For me current refers that there is actually in the DB. So what you call old.

I see you already have a $newProductIds below. I propose to change this one by $addedProductIds

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        $newProductIds = array_diff($currentProductIds, $oldProductIds);
+        $removedProductIds = array_diff($oldProductIds, $currentProductIds);
+
+        $productsToUpdate = [];
+        $group = $this->groupRepository->find($updateProductsToGroupCommand->groupId());
+
+        foreach ($newProductIds as $newProductId) {
+            $dbProduct = $this->productRepository->find($newProductId);
+            $dbProduct->addGroup($group);
+            $productsToUpdate[] = $dbProduct;
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php:45]: I don't know how many products we can have in a group, but if we add let's say 10.000 products, we would build an array of 10.000 products? It looks like a memory issue.

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/GroupUpdater.php [hunk 11]
```diff
@@ -92,12 +84,9 @@ protected function setData(GroupInterface $group, $field, $data)
             case 'type':
                 $this->setType($group, $data);
                 break;
-            case 'labels':
+                case 'labels':
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/GroupUpdater.php:87]: nope

File: tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php [hunk 12]
```diff
@@ -16,10 +16,11 @@
...
 {
     function let(
-        ObjectManager $objectManager,
-        EventDispatcherInterface $eventDispatcher,
+        ObjectManager                 $objectManager,
+        EventDispatcherInterface      $eventDispatcher,
         ProductUniqueDataSynchronizer $uniqueDataSynchronizer
-    ) {
+    )
+    {
```
Comment (nmarniesse) [inline tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php:23]: Please rollback this formatting and below

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsCommand.php [hunk 13]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+
+namespace Akeneo\Pim\Enrichment\Component\Product\Commands;
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsCommand.php:4]:
```suggestion
namespace Akeneo\Pim\Enrichment\Component\Product\Command;
```

Usually we don't pluralize the namespace

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsCommand.php [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+    /**
+     * @return array
+     */
+    public function productIds(): array
+    {
+        return $this->productIds;
+    }
+
+    public function __construct(string $groupId, array $uptodateProductIds)
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsCommand.php:31]: construct before other methods plz

**Review by nmarniesse — COMMENTED** [review]
Good job. Some comments nevertheless.  

Also there is a change in events. Before we update the group and we group the products before the POST_SAVE event. Now we update the group, send POST_SAVE, then group the products. But I don't see any solution and this should not be so important

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php (hunk 10 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php:45]: Yes, I'm aware of that...  The problem comes from the reuse of \Akeneo\Tool\Component\StorageUtils\Saver\BulkSaverInterface::saveAll
I inserted an additional warning  comment in header...

Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php:45]: I would call `saveAll()` each time the `$productsToUpdate` array reach 500 products for example. This way we would never have more than 500 products in memory

Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php:45]: Interesting point.
I agree this is the way it is done in the code everywhere: but It doesn't respect SoC principle... This kind of requirement  should be part of some reactive framework/api...
Moreover in the current context ( products are selected or deselected  through the UI it is not required...).
I suggest we make an improvement to investigate appropriate implementation.

File: tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php (hunk 12 above)
Comment (rfauglas) [reply tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php]: done

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsCommand.php (hunk 13 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsCommand.php]: done!

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php (hunk 10 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php]: I agree with both of you: while updating a lot of products could result in a OOM error again, it's little likely. I'd let the implem as is, and handle the case when/if the error occurs again.
That being said, IMHO, we should completely get rid of this screen (or leave it readonly), and only use the product grid and the 'Add to groups/Remove from groups' mass actions to handle this usecase

File: tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php [hunk 15]
```diff
@@ -34,11 +34,12 @@ function it_is_a_bulk_saver()
     }
 
     function it_saves_a_new_product(
-        ObjectManager $objectManager,
-        EventDispatcherInterface $eventDispatcher,
+        ObjectManager                 $objectManager,
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php:37]: please don't do that, it adds unnecessary changes to the diff IMO (especially when adding/removing an argument :slightly_smiling_face: )

File: tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php [hunk 16]
```diff
@@ -229,4 +236,15 @@ function it_does_not_save_multiple_products_if_none_was_updated(
...
+                                         EventDispatcherInterface      $eventDispatcher,
+                                         ProductUniqueDataSynchronizer $uniqueDataSynchronizer)
+    {
+        $uniqueDataSynchronizer->synchronize(Argument::any())->shouldNotBeCalled();
+        $eventDispatcher->dispatch(Argument::cetera())->shouldNotBeCalled();
+        $objectManager->persist(Argument::any())->shouldNotBeCalled();
+        $objectManager->flush()->shouldNotBeCalled();
+        $this->saveAll([]);
+    }
+}
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php:250]: don't forget to run php-cs-fixer :wink:

**Review by mmetayer — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php (hunk 10 above)
Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php]: The previous design was bad anyway: products are not part of the "Group" aggregate. There was no reason to dispatch an event "UpdatedGroup" when grouping a product. I'm not even sure that when when you group a product, the event "GroupUpated" is triggered :smile: 

Clearly, the design is the mirror of the UI design and its save button.

Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Component/Product/Commands/GroupProductsHandler.php]: > I don't know how many products we can have in a group, but if we add let's say 10.000 products, we would build an array of 10.000 products? It looks like a memory issue.

Not really, because it's only new products. And good luck to associate several thousands products with this screen...

**Review by ahocquard — APPROVED** [review]

File: tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php (hunk 15 above)
Comment (rfauglas) [reply tests/back/Pim/Enrichment/Specification/Bundle/Doctrine/Common/Saver/ProductSaverSpec.php:37]: sorry still have IDE formatting trouble I created an BH issue to tackle down...

**Review by nmarniesse — APPROVED** [review]
